[OSX] Improved pinch to zoom behaviour

alexanderweiss opened the pull request:
Improved the pinch to zoom behaviour on Macs with a trackpad:
- Reverse pinch to zoom to match the normal macOS behaviour: pinching out zooms in and pinching in zooms out
- Keep pinch to zoom working when setting the scroll-wheel to scroll the map rather than zoom. This is super useful because it allows users to use the trackpad both for scrolling and zooming. Note: disabling the scroll-wheel functionality entirely will still disable pinch to zoom.

alexanderweiss committed:
Change: [OSX] Reversed pinch to zoom behaviour.

Made pinch out zoom in and pinch in zoom out, as virtually all macOS
applications do.

michicc commented:
Code looks good, but somebody like @andythenorth should comment on expected OSX behaviour.

alexanderweiss committed:
Change: [OSX] Setting mouse-wheel to scroll the map does not disable pinch to zoom

alexanderweiss commented:
Okay!
Also updated the pull request as I'd forgotten the necessary special case for the map window.

andythenorth commented:
I will test this.  However my testing might be limited, as 'scroll-wheel-scrolls-map' is broken for me ;)  https://github.com/OpenTTD/OpenTTD/issues/6558

alexanderweiss commented:
Yes, I noticed that issue after having opened this pull request. I've made a fix for it, but I'm just checking that for issues. Would you like me to open a separate pull request for that, or just add it to this one @andythenorth?

Also, I noticed there's some unused logic in the MouseLoop function. Would it be appreciated if I cleaned it up a little?

EDIT: Fixed #6558 and made it feel a little more responsive

TrueBrain commented on src/window.cpp:
```diff
@@ -2901,19 +2897,28 @@ static void MouseLoop(MouseClick click, int mousewheel)
+				break;
 		}
 	}
+
```
It feels a bit weird that it depends if it was a `return` or `break` in the case above, especially as that is in an `if`. Sounds something that can break very easily over time.

Possibly it is better to put those lines in a function, and call that where you expect? (in 2 situations if I read it right? `MC_HOVER` and an `else` of the `if`?)

alexanderweiss replied:
I kind of agree. Initially I didn't return and had an if statement all the way at the end to exclude the specific situations where we shouldn't reset the values. However, the switch statement before this one (starting on [line 2870](https://github.com/alexanderweiss/OpenTTD/blob/8b6d2d470ef09e298acbc9358bab6d360ed494bf/src/window.cpp#L2870)) also uses early returns, so it seemed sensible to copy the pattern.

As I suggested in a comment on my pull request it may be valuable to actually clean up the whole MouseLoop function as there's more stuff that is a little weird. (E.g. the check for `w == NULL` on line 2903, which can never be true as there's an early return on line 2852 if `w == NULL`.) I'd be happy to do it, but I'm not sure if cleaning it up is appreciated?

TrueBrain replied:
Cleaning up can always be a good thing, but possibly not for this PR (maybe another?). But yeah, I understand why you did this. Okay, leave it as it is :) Now we just have to wait for @andythenorth  :D

TrueBrain commented:
@andythenorth the main question @michicc asked I guess, and at least which I have: is this expected behavior on Mac OS X, looking at other applications? :)

andythenorth commented:
Ok tested on macOS 10.12.6

- pinch-to-zoom confirmed now works per macOS conventions
- confirmed I can no longer trigger the issue in #6558 
- tested pinch-to-zoom with scrollwheel-scrolls-map and scrollwheel-zooms-map (that one is a weird setting anyway but eh, who asked my opinion? :) )

michicc commented:
The commits look good, but unfortunately can't be directly rebased as they conflict with a recently merged commit. I'd be grateful if you could update them.

Thanks for the much needed OSX love, BTW 🏅

alexanderweiss commented:
No problem. Just rebased it, @michicc. The changes in master introduced a bug breaking 2D scrolling when setting the viewport scrolling to use the left mouse buttons, so I've added one new commit to fix that.

And thanks for checking @andythenorth!

PeterN commented:
Your fix for scrollwheel scrolling should be separate, not added on here. I'm not convinced it's correct either as it doesn't test for WF_DISABLE_VP_SCROLL.

alexanderweiss commented:
I’ll create a second pull request then. And yes, you’re right about WF_DISABLE_VP_SCROLL. Initially added it, then removed it unintentionally.

> On 30 Apr 2018, at 22:42, PeterN <notifications@github.com> wrote:
> 
> Your fix for scrollwheel scrolling should be separate, not added on here. I'm not convinced it's correct either as it doesn't test for WF_DISABLE_VP_SCROLL.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/OpenTTD/OpenTTD/pull/6753#issuecomment-385522334>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAJLfGL9dT9uYg1hQhVypnQBxvwHURPCks5tt3ccgaJpZM4Tq8zj>.
>

PeterN commented:
The current implementation of mouse-wheel scrolling is pretty broken anyway. I have a better implementation work in progress, but I can't test the OS X variant. Anyway, this should be a separate issue :-)

LordAro commented:
@alexanderweiss are you likely to find some time to split the scrollwheel scrolling change out? We'd like to get the bit we like merged in sooner rather than later :)

alexanderweiss commented:
@LordAro I removed the 2D scrolling fixes and improvements from this pull request (including the fix for #6558) so it's just reversing pinch to zoom & keeping it enabled when setting 2D scrolling to scroll the map. I'll create a second pull request for the 2D scrolling improvements.
